src/parser/sparqlParser/SparqlQleverVisitor.h

```diff
@@ -1070,16 +1113,46 @@ class SparqlQleverVisitor : public SparqlAutomaticVisitor {
 
   antlrcpp::Any visitIriOrFunction(
       SparqlAutomaticParser::IriOrFunctionContext* ctx) override {
-    if (ctx->argList()) {
-      throw SparqlParseException{
-          "calls to non-built-in functions in expressions are not supported by "
-          "this parser"};
+    // Case 1: Just an IRI.
+    if (ctx->argList() == nullptr) {
+      return ExpressionPtr{
+          std::make_unique<sparqlExpression::StringOrIriExpression>(
+              ctx->getText())};
     }
-
-    return ExpressionPtr{
-        std::make_unique<sparqlExpression::StringOrIriExpression>(
-            ctx->getText())};
-    return visitChildren(ctx);
+    // Case 2: Function call, where the function name is an IRI.
+    auto iri = std::move(visitIri(ctx->iri()).as<std::string>());
+    auto argList = std::move(
+        visitArgList(ctx->argList()).as<std::vector<ExpressionPtr>>());
+    const char* geofPrefix = "<http://www.opengis.net/def/function/geosparql/";
+    if (iri.rfind("<http://www.opengis.net/def/function/geosparql/", 0) == 0) {
+      size_t pos = strlen(geofPrefix);
+      if (iri.compare(pos, std::string::npos, "distance>") == 0) {
+        if (argList.size() == 2) {
+          return createExpression<sparqlExpression::DistExpression>(
+              std::move(argList[0]), std::move(argList[1]));
+        } else {
+          throw SparqlParseException{
+              "Function geof:distance requires two arguments"};
+        }
+      } else if (iri.compare(pos, std::string::npos, "longitude>") == 0) {
+        if (argList.size() == 1) {
+          return createExpression<sparqlExpression::LongitudeExpression>(
+              std::move(argList[0]));
+        } else {
+          throw SparqlParseException{
+              "Function geof:longitude requires one argument"};
+        }
+      } else if (iri.compare(pos, std::string::npos, "latitude>") == 0) {
+        if (argList.size() == 1) {
+          return createExpression<sparqlExpression::LatitudeExpression>(
+              std::move(argList[0]));
+        } else {
+          throw SparqlParseException{
+              "Function geof:longitude requires one argument"};
+        }
+      }
+    }
+    throw SparqlParseException{"Function \"" + iri + "\" not supported"};
```

Comment:
There is a lot of repeated boilerplate code here, 
we could also think about making this more generic, but we can discuss this together, and it is not required for this PR, but once we have 100 custom functions it would probably be nice:)

Reply:
I agree!